0024-pre-shared-key-v2/0024-pre-shared-key-v2.md

```diff
@@ -0,0 +1,287 @@
+---
+title: Pre-Shared Key V2 (PSKv2) Transport Protocol
+draft: 1
+---
+# Pre-Shared Key V2 (PSKv2) Transport Protocol
+
+This spec defines version 2 of the Pre-Shared Key Transport Protocol. PSKv2 implements end-to-end data encryption, condition and fulfillment generation, [End-to-End Quoting](#end-to-end-quoting) and [Chunked Payments](#chunked-payments). The approach to chunked payments is inspired by the Internet's Transmission Control Protocol (TCP).
+
+**Table of Contents**
+
+* [Overview](#overview)
+  + [Setup](#setup)
+  + [Encryption](#encryption)
+  + [Condition and Fulfillment Generation](#condition-and-fulfillment-generation)
+  + [End-To-End Quoting](#end-to-end-quoting)
+  + [Chunked Payments](#chunked-payments)
+* [Model of Operation](#model-of-operation)
+  + [Informational Quote](#informational-quote)
+  + [Single Payment](#single-payment)
+  + [Chunked Payment](#chunked-payment)
+* [Specification](#specification)
+  + [Encryption](#encryption-1)
+  + [Data Packet](#data-packet)
+* [Appendix A: Advanced Sender Features](#appendix-a-advanced-sender-features)
+
+## Overview
+
+Like [PSKv1](./0016-pre-shared-key/0016-pre-shared-key.md), PSKv2 uses a secret shared between the sender and receiver to generate the conditions and fulfillments for payments without repeated end-to-end communication. PSKv2 additionally includes support for End-to-End Quoting and Chunked Payments.
+
+This specificiation aims to standardize the behavior of receivers while allowing for sophisticated sender behavior to be developed in the future. [Appendix A: Advanced Sender Features](#appendix-a-advanced-sender-features) outlines areas for potential sender features to be implemented in the future.
+
+### Setup
+
+Before using PSKv2, the sender and receiver use an authenticated, encrypted communication channel (generally provided as part of an [Application Layer Protocol](./0001-interledger-architecture/0001-interledger-architecture.md#application-layer)) to exchange:
+
+- PSK version (such as `2.0` for this version)
+- A 32-byte random or pseudorandom shared secret
+- The receiver's ILP address
+
+The receiver MAY generate the shared secret from a longer-term secret and a nonce, as described in [PSKv1's Appendix A: Recommended Algorithm for Generating Shared Secrets](./0016-pre-shared-key/0016-pre-shared-key.md#appendix-a-recommended-algorithm-for-generating-shared-secrets). In this case, the receiver would append the nonce (**not** the secret) to their ILP address to create the destination address they communicate to the sender.
+
+### Encryption
+
+In PSKv2.0, request and response data is encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 12-byte Initialization Vector (IV) and a 16-Byte Authentication Tag.
```

Comment:
From what source did you determine that those were the correct lengths to use? (I think PSK v1 uses a 16-byte IV)

Reply:
https://github.com/interledger/rfcs/pull/351#discussion_r156799912